src/BlockTypes/ProductQuery.php

```diff
@@ -186,16 +177,43 @@ function( $acc, $query ) {
 	/**
 	 * Merge in the first parameter the keys "post_in", "meta_query" and "tax_query" of the second parameter.
 	 *
-	 * @param array $a The first query.
-	 * @param array $b The second query.
+	 * @param array[] ...$queries Query arrays to be merged.
 	 * @return array
 	 */
-	private function merge_queries( $a, $b ) {
-		$a['post__in']   = ( isset( $b['post__in'] ) && ! empty( $b['post__in'] ) ) ? $this->intersect_arrays_when_not_empty( $a['post__in'], $b['post__in'] ) : $a['post__in'];
-		$a['meta_query'] = ( isset( $b['meta_query'] ) && ! empty( $b['meta_query'] ) ) ? array_merge( $a['meta_query'], array( $b['meta_query'] ) ) : $a['meta_query'];
-		$a['tax_query']  = ( isset( $b['tax_query'] ) && ! empty( $b['tax_query'] ) ) ? array_merge( $a['tax_query'], array( $b['tax_query'] ) ) : $a['tax_query'];
+	private function merge_queries( ...$queries ) {
+		$merged_query = array_reduce(
+			$queries,
+			function( $acc, $query ) {
+				if ( ! is_array( $query ) ) {
+					return $acc;
+				}
+				// If the $query doesn't contain any valid query keys, we unpack it then merge.
+				if ( empty( array_intersect( $this->get_valid_query_vars(), array_keys( $query ) ) ) ) {
+					return $this->merge_queries( $acc, ...array_values( $query ) );
+				}
+				return $this->array_merge_recursive_replace_non_array_properties( $acc, $query );
```

Comment:
Would you mind explaining a little bit more about this logic?

1. What do you mean we “unpack it”?
2. What are the use-cases for non valid query keys? Are they our custom `__woocommerce` queries?
3. Why do we recursively replace them?

I'd just like to understand the entire flow a bit more and why these things are needed (and perhaps it could be useful to document them in the PHPDocs for posterity).

Reply:
> 1. What do you mean we “unpack it”?

It's restructuring. I used unpack as PHP uses that word. I updated the comment to include both (`unpack/destructure`).

> 3. Why do we recursively replace them?

We recursively replace only non-array values. `array_merge_recursive()`, which I used initially, will merge the value with the same key, but it's not what we want. For example:
```
$base = ['orderby' => 'date']
$new  = ['orderby' => 'meta_value_num']

With array_merge_recursive(), the result will be: ['orderby' => ['date', 'meta_value_num']]
But we want this: ['orderby' => 'meta_value_num']

```

> 2. What are the use-cases for non valid query keys? Are they our custom __woocommerce queries?

_Explaining this makes me think I might be overthinking._

> Are they our custom __woocommerce queries?

No. Query keys (or query vars) here are WP_Query variables. Using valid here is misleading, it should be `built-in` instead. I check if the current array contains the built-in query vars to deal with unknow incoming arrays. For example:

If we feed `merge_queries` with an array like:

```
[
	[
		'meta_query' => [
			...<meta_queries_here>
		]
	]
]
```
`merge_queries` will restructure the array until it gets an array containing the `meta_query` key, which is a built-in query variable, then starts merging.

But, as this is a private method, which is only used inside the ProductQuery class, so we can control the inputs of `merge_queries` calls, I think this logic can be simplified by destructuring until we get the associate arrays, new commit is coming...

Reply:
> But, as this is a private method, which is only used inside the ProductQuery class, so we can control the inputs of `merge_queries` calls, I think this logic can be simplified by destructuring until we get the associative arrays, new commit is coming...

We can't do this because we're returning associative arrays, for example in `get_queries_by_attributes()`.

Reply:
After considering, I reverted e6df617. I think the benefit of a versatile `merge_queries()` method outperforms the complexity checking for valid query variables.

Reply:
Thanks a lot @dinhtungdu for the amazing explanation and in-depth commentary! This was extremely helpful to understand this somewhat convoluted logic. And I agree that simplifying is better in this case: glad that some rubberducking was also useful to you.

Just a small note:

>> What do you mean we “unpack it”?

> It's restructuring. I used unpack as PHP uses that word. I updated the comment to include both (unpack/destructure).

By this you mean: `...array_values( $query )` right? In which case, I guess we call this “unpacking” “spreading” and not “destructuring” in the JS world:

> Destructuring is used to create varibles from array items or object properties. Spread syntax is used to unpack iterables such as arrays, objects, and function calls.

Just in case you want to update your comment. But it's not a big deal. Thanks a lot for unpacking ( 🤭 ) the code for me.

Reply:
> By this you mean: `...array_values( $query )` right? In which case, I guess we call this “unpacking” “spreading” and not “destructuring” in the JS world:

> Just in case you want to update your comment. But it's not a big deal. Thanks a lot for unpacking ( 🤭 ) the code for me.

I usually misuse those terms 🤦🏽. I updated the code. Thanks so much, it's easier to memorize for me now 🥰